kdc/kerberos5.c

```diff
@@ -2046,10 +2048,179 @@ get_local_tgs(krb5_context context,
     return ret;
 }
 
-/*
- *
- */
+static int
+siva_any_to_certs(hx509_context context, const SignedData *sd, hx509_certs certs)
+{
+    int ret;
+    size_t i;
 
+    if (sd->certificates == NULL)
+	return 0;
+
+    for (i = 0; i < sd->certificates->len; i++) {
+	heim_error_t error;
+	hx509_cert c;
+
+	c = hx509_cert_init_data(context,
+				 sd->certificates->val[i].data,
+				 sd->certificates->val[i].length,
+				 &error);
+	if (c == NULL) {
+	    ret = heim_error_get_code(error);
+	    heim_release(error);
+	    return ret;
+	}
+	ret = hx509_certs_add(context, certs, c);
+	hx509_cert_free(c);
+	if (ret)
+	    return ret;
+    }
+
+    return 0;
+}
+
+static int siva_extract_san_from_padata(astgs_request_t r, PrincipalName *princname) {
```

Comment:
Let's add a block comment here to explain why this code exists at all.

Also, do please run the KDC with `valgrind` or other memory debugger to make sure there's no memory errors here when exercising this code.

Ah, we do need to write a test.  Hmmm, to write a test we'll need a way to generate an `AS-REQ` with an X.500 `cname`.  That does make this more involved...  We'll need to add X.500 support in `lib/krb5/principal.c`, and then we'd need a command-line option to `kuser/kinit.c` to indicate that we want an X.500 style client principal name.

Reply:
Most probably, I will be removing this code, since the test I'm doing doesn't seem to require it. I'll update the draft PR with what I have so far, but may need further help in debugging why exactly Windows doesn't consider the successful AS-REP "a successful authentication attempt".